nx/lib/nx/random.ex

```diff
@@ -0,0 +1,419 @@
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [min_val, max_val).
+  
+  ## Options
+  
+    * `:type` - the integer type for the returned tensor
+    * `:shape` - shape of the returned tensor
+  
+  
+  """
+  defn randint(key, min_val, max_val, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    assert_key(key)
+
+    shape = opts[:shape]
+    type = {_, nbits} = normalize(opts[:type])
+
+    case type do
+      {:u, _} ->
+        :ok
+
+      {:s, _} ->
+        :ok
+
+      _ ->
+        raise ArgumentError,
+              "expected integer type, got type #{inspect(type)}"
+    end
+
+    keys = split(key)
+
+    min_val = Nx.broadcast(min_val, shape)
+    max_val = Nx.broadcast(max_val, shape)
+
+    higher_bits = random_bits(keys[0], shape: shape, bit_width: nbits)
+    lower_bits = random_bits(keys[1], shape: shape, bit_width: nbits)
+    span = max_val - min_val
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits
+      |> Nx.remainder(span)
+      |> Nx.multiply(multiplier)
+      |> Nx.add(Nx.remainder(lower_bits, span))
+      |> Nx.remainder(span)
+
+    Nx.as_type(min_val + offset, type)
+  end
+
+  @doc """
+  Sample uniform float values in [min_val, max_val).
+  
+  ## Options
+  
+    * `:type` - a float type for the returned tensor
+  
+    * `:shape` - shape of the returned tensor
+  
+    * `:min_val` - minimum value, default is 0.0
+  
+    * `:max_val` - maximum value, default is 1.0
+  
+  """
+  defn uniform(key, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:f, 32}, min_val: 0.0, max_val: 1.0)
+
+    assert_key(key)
+
+    type = {_type, nbits} = normalize(opts[:type])
+
+    case type do
+      {:f, _} ->
+        :ok
+
+      {:bf, _} ->
+        :ok
+
+      _ ->
+        raise ArgumentError,
+              "expected float type, got type #{inspect(type)}"
+    end
+
+    info = float_info(type)
+
+    shape = opts[:shape]
+    min_val = opts[:min_val]
+    max_val = opts[:max_val]
+
+    u_one = Nx.tensor(1.0, type: type) |> Nx.bitcast({:u, nbits})
+
+    random_bits(key, shape: shape, bit_width: nbits)
+    |> Nx.as_type({:u, nbits})
+    |> Nx.right_shift(Nx.tensor(nbits - info[:mantissa], type: {:u, nbits}))
+    |> Nx.bitwise_or(u_one)
+    |> Nx.bitcast(type)
+    |> Nx.subtract(Nx.tensor(1.0, type: type))
+    |> Nx.multiply(max_val - min_val)
+    |> Nx.add(min_val)
+    |> Nx.reshape(shape)
+    |> Nx.max(min_val)
+  end
+
+  deftransformp normalize(type), do: Nx.Type.normalize!(type)
+
+  defnp assert_key(tensor) do
+    %{shape: shape, type: type} = tensor
+
+    case shape do
+      {2} ->
+        :ok
+
+      _ ->
+        raise(
+          ArgumentError,
+          "expected key to have shape {2}, got tensor with shape #{inspect(shape)}"
+        )
+    end
+
+    case type do
+      {:u, _} ->
+        :ok
```

Comment:
Should we assert the key is explicitly u32, since that's how it is defined? For example, would the algorithms still work if the keys are s64 or u16?

Reply:
Yes, key has to be explicitly u32